language/move-lang/src/shared/ast_debug.rs

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pub struct AstWriter {
+    verbose: bool,
+    margin: usize,
+    lines: Vec<String>,
+}
+
+impl AstWriter {
+    fn new(verbose: bool) -> Self {
+        Self {
+            verbose,
+            margin: 0,
+            lines: vec![String::new()],
+        }
+    }
+
+    fn normal() -> Self {
+        Self::new(false)
+    }
+
+    fn verbose() -> Self {
+        Self::new(true)
+    }
+
+    fn cur(&mut self) -> &mut String {
+        self.lines.last_mut().unwrap()
+    }
+
+    pub fn new_line(&mut self) {
+        self.lines.push(String::new());
+    }
+
+    pub fn write(&mut self, s: &str) {
+        let margin = self.margin;
+        let cur = self.cur();
+        if cur.is_empty() {
+            (0..margin).for_each(|_| cur.push(' '));
+        }
+        cur.push_str(s);
+    }
+
+    pub fn writeln(&mut self, s: &str) {
+        self.write(s);
+        self.new_line();
+    }
+
+    pub fn indent<F: FnOnce(&mut AstWriter) -> ()>(&mut self, inc: usize, f: F) {
+        self.new_line();
+        self.margin += inc;
+        f(self);
+        self.margin -= inc;
+        self.new_line();
+    }
+
+    pub fn block<F: FnOnce(&mut AstWriter) -> ()>(&mut self, f: F) {
+        self.write(" {");
+        self.indent(4, f);
+        self.write("}");
+    }
+
+    pub fn annotate<F: FnOnce(&mut AstWriter) -> (), Annot: AstDebug>(
+        &mut self,
+        f: F,
+        annot: &Annot,
+    ) {
+        if self.verbose {
+            self.write("(");
+        }
+        f(self);
+        if self.verbose {
+            self.write(": ");
+            annot.ast_debug(self);
+            self.write(")");
+        }
+    }
+
+    pub fn list<T, F: FnMut(&mut AstWriter, T) -> bool>(
+        &mut self,
+        items: impl std::iter::IntoIterator<Item = T>,
+        sep: &str,
+        mut f: F,
+    ) {
+        let iter = items.into_iter();
+        let len = match iter.size_hint() {
+            (lower, None) => {
+                assert!(lower == 0);
+                return;
+            }
+            (_, Some(len)) => len,
+        };
+        for (idx, item) in iter.enumerate() {
+            let needs_newline = f(self, item);
+            if idx + 1 != len {
+                self.write(sep);
+                if needs_newline {
+                    self.new_line()
+                }
+            }
+        }
+    }
+
+    pub fn comma<T, F: FnMut(&mut AstWriter, T) -> ()>(
+        &mut self,
+        items: impl std::iter::IntoIterator<Item = T>,
+        mut f: F,
+    ) {
+        self.list(items, ", ", |w, item| {
+            f(w, item);
+            false
+        })
+    }
+
+    pub fn semicolon<T, F: FnMut(&mut AstWriter, T) -> ()>(
+        &mut self,
+        items: impl std::iter::IntoIterator<Item = T>,
+        mut f: F,
+    ) {
+        self.list(items, ";", |w, item| {
+            f(w, item);
+            true
+        })
+    }
+}
+
+pub trait AstDebug {
```

Comment:
Thanks! Could you add documentation about those two things to this trait?

Reply:
About which two?

Reply:
Just the two things you mentioned :)

> Hopefully, you shouldn't have to remember. I tried to set it up so it yells at you if a new field is added. At the beginning of each declaration, I fully pattern match against the struct. So it will yell if a field is missing (no ..s are used)

> The deriving is something I looked into, but its not really easy to do, as each field does not get printed in a direct manner, and most of the logic is ad hoc

They're both great answers to reasonable questions and should be in an "implementation notes" section in a doc comment on the trait.